vm/actor/src/lib.rs

```diff
@@ -36,17 +41,91 @@ fn check_empty_params(params: &Serialized) -> Result<(), EncodingError> {
 
 /// Create a map
 #[inline]
-fn make_map<BS: BlockStore>(store: &'_ BS) -> Hamt<'_, String, BS> {
+fn make_map<BS: BlockStore>(store: &'_ BS) -> Hamt<'_, BytesKey, BS> {
     Hamt::new_with_bit_width(store, HAMT_BIT_WIDTH)
 }
 
-pub fn deal_key(d: DealID) -> String {
+pub fn deal_key(d: DealID) -> BytesKey {
     let mut bz = unsigned_varint::encode::u64_buffer();
     unsigned_varint::encode::u64(d, &mut bz);
-    String::from_utf8_lossy(&bz).to_string()
+    bz.to_vec().into()
 }
 
-pub fn parse_uint_key(s: &str) -> Result<u64, UVarintError> {
-    let (v, _) = unsigned_varint::decode::u64(s.as_ref())?;
+pub fn parse_uint_key(s: &[u8]) -> Result<u64, UVarintError> {
+    let (v, _) = unsigned_varint::decode::u64(s)?;
     Ok(v)
 }
+
+/// Key type to be used to isolate usage of unsafe code and allow non utf-8 bytes to be
+/// serialized as a string.
+// TODO revisit to change serialization in spec or find a non-unsafe way to do this
+#[derive(Eq, PartialOrd, Clone, Debug)]
+pub struct BytesKey(pub Vec<u8>);
+
+impl PartialEq for BytesKey {
+    fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0
+    }
+}
+
+impl Hash for BytesKey {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        state.write(&self.0);
+    }
+}
+
+impl Borrow<[u8]> for BytesKey {
+    fn borrow(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl Borrow<Vec<u8>> for BytesKey {
+    fn borrow(&self) -> &Vec<u8> {
+        &self.0
+    }
+}
+
+impl Deref for BytesKey {
+    type Target = Vec<u8>;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Serialize for BytesKey {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        unsafe { str::from_utf8_unchecked(&self.0) }.serialize(serializer)
```

Comment:
Something sounds wrong here, quoting from the Cbor RFC:

> a text string, specifically a string of Unicode characters that is encoded as UTF-8

So either the go implementation is generating invalid cbor, or storing its strings as bytes. If they do use bytes, you can just use bytes as well and avoid the unsafe. If they generate invalid cbor, that should be fixed and not worked around.

Reply:
ah yeah, I mentioned over a call to others that this was a temporary measure and I indicate it in the TODO, I will bring it up as an issue in the go implementations, since they are encoding Hamt keys as strings when they are actually non utf8 bytes. I hate having to do this, but need compatibility for now until this is resolved